pallets/api/src/nonfungibles/mod.rs

```diff
@@ -0,0 +1,591 @@
+				)
+				.map_err(|e| {
+					e.with_weight(NftsWeightInfoOf::<T>::approve_transfer(item.is_some() as u32))
+				})?;
+				NftsWeightInfoOf::<T>::approve_transfer(item.is_some() as u32)
+			} else {
+				NftsOf::<T>::cancel_approval(
+					origin,
+					collection,
+					item,
+					T::Lookup::unlookup(operator.clone()),
+				)
+				.map_err(|e| {
+					e.with_weight(NftsWeightInfoOf::<T>::cancel_approval(item.is_some() as u32))
+				})?;
+				NftsWeightInfoOf::<T>::cancel_approval(item.is_some() as u32)
+			};
+			Self::deposit_event(Event::Approval { collection, item, operator, owner, approved });
+			Ok(Some(weight).into())
+		}
+
+		/// Issue a new collection of non-fungible items from a public origin.
+		///
+		/// # Parameters
+		/// - `admin` - The admin of this collection. The admin is the initial address of each
+		/// member of the collection's admin team.
+		/// - `config` - The configuration of the collection.
+		#[pallet::call_index(7)]
+		#[pallet::weight(NftsWeightInfoOf::<T>::create())]
+		pub fn create(
+			origin: OriginFor<T>,
+			admin: AccountIdOf<T>,
+			config: CollectionConfigFor<T>,
+		) -> DispatchResult {
+			let creator = ensure_signed(origin.clone())?;
+			// TODO: re-evaluate next collection id in nfts pallet. The `Incrementable` trait causes
+			//  issues for setting it to xcm's `Location`. This can easily be done differently.
+			let id = NextCollectionIdOf::<T>::get()
+				.or(T::CollectionId::initial_value())
+				.ok_or(NftsErrorOf::<T>::UnknownCollection)?;
+			NftsOf::<T>::create(origin, T::Lookup::unlookup(admin.clone()), config)?;
+			Self::deposit_event(Event::Created { id, creator, admin });
+			Ok(())
+		}
+
+		/// Destroy a collection of fungible items.
+		///
+		/// # Parameters
+		/// - `collection` - The collection to destroy.
+		/// - `witness` - Information on the items minted in the collection. This must be
+		/// correct.
+		#[pallet::call_index(8)]
+		#[pallet::weight(NftsWeightInfoOf::<T>::destroy(
+    		witness.item_metadatas,
+    		witness.item_configs,
+    		witness.attributes,
+            witness.item_holders,
+            witness.allowances,
+		))]
+		pub fn destroy(
+			origin: OriginFor<T>,
+			collection: CollectionIdOf<T>,
+			witness: DestroyWitness,
+		) -> DispatchResultWithPostInfo {
+			NftsOf::<T>::destroy(origin, collection, witness)
+		}
+
+		/// Set an attribute for a collection or item.
+		///
+		/// # Parameters
+		/// - `collection` - The collection whose item's metadata to set.
+		/// - `maybe_item` - The item whose metadata to set.
+		/// - `namespace` - Attribute's namespace.
+		/// - `key` - The key of the attribute.
+		/// - `value` - The value to which to set the attribute.
+		#[pallet::call_index(12)]
+		#[pallet::weight(NftsWeightInfoOf::<T>::set_attribute())]
+		pub fn set_attribute(
+			origin: OriginFor<T>,
+			collection: CollectionIdOf<T>,
+			item: Option<ItemIdOf<T>>,
+			namespace: AttributeNamespaceOf<T>,
+			key: BoundedVec<u8, T::KeyLimit>,
```

Comment:
Could this parameter be changed to `Vec<u8>` and we handle the length check in the runtime, so contract api implementation does not need to handle it?